### PR TITLE
feat(hooks): redirect Asana URL WebFetch to the asana CLI

### PR DESCRIPTION
**Pull request by amondnet**
## Summary

Adds a `PreToolUse` hook that intercepts `WebFetch` and `Fetch` tool calls targeting `app.asana.com` URLs and redirects the agent to the local `asana` CLI instead of hitting the Asana web app.

The hook covers both URL forms used by the Asana web app:
- **V0**: `https://app.asana.com/0/<project>/<task>`
- **V1**: `https://app.asana.com/1/...`

When a match is detected, the hook blocks the fetch and returns a structured error that names the correct CLI commands:
- `asana task get <task-id>` — read a task
- `asana project get <project-id>` — read a project
- `asana task comment list <task-id> --format toon` — list comments

## Changes

- `hooks/intercept-webfetch.ts` — PreToolUse hook implementation
- `hooks/hooks.json` — hook registration (wires the hook into Claude Code's PreToolUse pipeline)
- `test/hooks/intercept-webfetch.test.ts` — bun:test suite (14 tests covering V0/V1 URL matching, non-Asana pass-through, and CLI hint formatting)

## Verification

- `bun test test/hooks/intercept-webfetch.test.ts` → 14 tests pass
- `bun run lint` → clean
- `bunx tsc --noEmit` → no new errors

## Notes

Adapted from the `chatbot-pf/engineering-standards` MCP-variant hook to target this repo's CLI. Tests were ported from the original vitest suite to bun:test.

**Intentionally excluded from this PR:**
- `.claude/settings.json` (a separate plugin-enablement change — nostics/axi — unrelated to this hook)
- `.idea/`, `.agents/`, `.claude/agent-memory/`, `.claude/skills/`, `.please/docs/review/`, `skills-lock.json` (local/untracked artifacts)

---
## Summary by cubic
Blocks `WebFetch`/`Fetch` calls to `app.asana.com` and redirects to the local `asana` CLI for authenticated, structured access. Supports V0 and V1 links (task, project, comment) and preserves Asana task focus URLs (`/f`) via intentional prefix matching.

- **New Features**
  - Added `PreToolUse` hook (`hooks/intercept-webfetch.ts`) and registration (`hooks/hooks.json`).
  - Denies Asana fetches with guidance for:
    - `asana task get <task-id> --format toon`
    - `asana project get <project-id> --format toon`
    - `asana task comment list <task-id> --format toon`
  - Non-Asana URLs pass through; on hook error, fetch proceeds with a warning.
  - Tests: cover V0/V1, host boundary, both tool paths, CLI guidance, and lenient task matching for `/f` focus URLs with comment-route precedence.

- **Bug Fixes**
  - Strict hostname check using `new URL()` to avoid look-alike/proxied domains.
  - Pathname matching with anchored `RegExp.exec()` for accuracy.
  - Extracted pure `decideForToolCall()` for unit tests; deny message uses a warning prefix.

<sup>Written for commit de95547184b92adc77092a4800a672c828452cfe. Summary will update on new commits.</sup>

## Summary by CodeRabbit

* **New Features**
  * Added a pre-tool hook that intercepts web fetches targeting supported Asana links and instructs you to use the local CLI instead.
  * Recognizes both legacy and versioned Asana URL patterns (tasks, projects, and comments) and passes through non-Asana or unrecognized inputs without blocking.
* **Tests**
  * Expanded test coverage for Asana URL detection, generated CLI guidance, and tool-call permission decisions (including edge cases and allow/pass-through behavior).

### CODE REVIEW

File: hooks/intercept-webfetch.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+  // V0 (legacy): Project - /0/{project_id}
+  const v0ProjectMatch = url.match(/app\.asana\.com\/0\/(\d+)\/?(?:\?|$)/)
+  if (v0ProjectMatch) {
+    const [, projectId] = v0ProjectMatch
+    return { type: 'project', projectId }
+  }
+
+  return null
+}
```
[the remark this answers is not in the document]
Comment (amondnet) [reply hooks/intercept-webfetch.ts]: @gemini-code-assist ✅ 적용했습니다. `new URL()`로 파싱 후 `hostname === 'app.asana.com'`을 검증하고 pathname에 대해서만 매칭하도록 변경했습니다. `app.asana.com.evil.com` 및 proxy 쿼리 케이스 회귀 테스트도 추가했습니다 (c649664). 비문자열·malformed 입력은 `new URL()`의 try/catch가 모두 null로 처리하므로 별도 typeof 검사는 생략했습니다.

File: hooks/intercept-webfetch.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    const [, projectId, taskId] = v0TaskMatch
+    return { type: 'task', projectId, taskId }
+  }
+
+  // V0 (legacy): Project - /0/{project_id}
+  const v0ProjectMatch = url.match(/app\.asana\.com\/0\/(\d+)\/?(?:\?|$)/)
+  if (v0ProjectMatch) {
+    const [, projectId] = v0ProjectMatch
+    return { type: 'project', projectId }
+  }
```
[the remark this answers is not in the document]
Comment (amondnet) [reply hooks/intercept-webfetch.ts]: @coderabbitai ✅ host 경계를 `new URL()` hostname 검증으로 앵커링했고, 매칭을 `RegExp.exec()` + anchored pathname 패턴으로 바꿔 SonarCloud `prefer-regexp-exec` 경고도 함께 해소했습니다 (c649664).

File: hooks/intercept-webfetch.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    const hookInput: PreToolUseHookInput = JSON.parse(input)
+
+    // Only act on fetch-style tools; pass everything else through untouched.
+    if (hookInput.tool_name !== 'WebFetch' && hookInput.tool_name !== 'Fetch') {
+      process.stdout.write(`${JSON.stringify({})}\n`)
+      process.exit(0)
+    }
+
+    const url = (hookInput.tool_input as WebFetchToolInput)?.url
+    const match = url ? parseAsanaUrl(url) : null
```
[the remark this answers is not in the document]
Comment (amondnet) [reply hooks/intercept-webfetch.ts]: @greptile-apps ✅ `decideForToolCall()`를 순수 함수로 추출해 `Fetch` tool_name 경로를 단위 테스트로 커버했습니다 — WebFetch와 동일하게 deny되는지, 그리고 url 필드가 없는 Fetch는 pass-through 되는지 검증합니다 (c649664).

File: hooks/intercept-webfetch.ts [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+      process.exit(0)
+    }
+
+    const output: HookJSONOutput = {
+      hookSpecificOutput: {
+        hookEventName: 'PreToolUse',
+        permissionDecision: 'deny',
+        permissionDecisionReason: 'Asana URL detected — use the asana CLI for authenticated, structured access',
+      },
+      systemMessage: `✓ ${buildCliGuidance(match)}`,
```
[the remark this answers is not in the document]
Comment (amondnet) [reply hooks/intercept-webfetch.ts]: @greptile-apps ✅ deny 메시지 prefix를 `✓`에서 `⚠️`로 변경했습니다. `✓` 미사용을 고정하는 테스트도 추가했습니다 (c649664).

File: hooks/intercept-webfetch.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (amondnet) [reply hooks/intercept-webfetch.ts]: @greptile-apps ✅ `new URL()` hostname 검증 + pathname 매칭으로 origin 기반으로 전환했습니다. fragment(`#section`)도 pathname이 제외하므로 함께 처리됩니다 (c649664).

File: hooks/intercept-webfetch.ts [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+  commentId?: string
+}
+
+/**
+ * Parse an Asana URL and extract its identifiers, or return null if it is not
+ * a recognizable task/project/comment URL.
+ */
+export function parseAsanaUrl(url: string): AsanaUrlMatch | null {
+  // V1: Task with comment - /task/{task_id}/comment/{comment_id}
+  const v1CommentMatch = url.match(/app\.asana\.com\/1\/(\d+)\/(?:project\/(\d+)\/)?task\/(\d+)\/comment\/(\d+)/)
```
[the remark this answers is not in the document]
Comment (amondnet) [reply hooks/intercept-webfetch.ts]: @cubic-dev-ai ✅ substring 매칭을 `new URL()` hostname 검증으로 교체했습니다. 비-Asana origin에 `app.asana.com` 문자열이 포함돼도 더 이상 deny되지 않습니다 (c649664).

File: hooks/intercept-webfetch.ts [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+  // V1: Project - /1/{workspace}/project/{project}
+  const v1Project = /^\/1\/(\d+)\/project\/(\d+)\/?$/.exec(path)
+  if (v1Project) {
+    const [, workspaceId, projectId] = v1Project
+    return { type: 'project', workspaceId, projectId }
+  }
+
+  // V0 (legacy): Task - /0/{project}/{task}
+  const v0Task = /^\/0\/(\d+)\/(\d+)/.exec(path)
```
[the remark this answers is not in the document]
Comment (amondnet) [reply hooks/intercept-webfetch.ts:86]: @coderabbitai 이 제안은 의도적으로 받지 않았습니다(declined). Asana는 실제 task URL에 focus 접미사 `/f`를 붙입니다 — 예: `app.asana.com/0/{project}/{task}/f`, `.../task/{task}/f`. 제안대로 task/comment 패턴을 `\/?$`로 정확 매칭하면 이 흔한 focus URL들이 매칭에서 빠져 hook이 실제 task URL 인터셉트를 놓치게 됩니다(인터셉션 커버리지 회귀).

project 라우트는 이미 `$`로 정확 앵커링되어 있고, comment 라우트는 task보다 먼저 검사되어 절대 가려지지 않습니다. 따라서 task/comment의 prefix 매칭은 의도된 동작입니다.

이 의도를 회귀 테스트로 고정했습니다 (de95547): V0/V1 `/f` URL이 task로 라우팅되는지 + comment 우선순위. 향후 누군가 '정확 매칭으로 강화'하면 테스트가 실패해 막아줍니다.

(참고: Tools 섹션의 OpenGrep `command-injection.exec-js`는 오탐입니다 — `RegExp.prototype.exec`를 `child_process.exec`로 오인한 것으로, 자식 프로세스 호출이 없습니다.)

File: hooks/intercept-webfetch.ts [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+const ASANA_HOST = 'app.asana.com'
+
+/**
+ * Match the Asana identifiers in a URL *pathname* (host already verified).
+ * Pathname excludes query/fragment, so anchored `^…$` patterns are exact.
+ */
+function matchAsanaPath(path: string): AsanaUrlMatch | null {
+  // V1: Task with comment - /1/{workspace}/[project/{project}/]task/{task}/comment/{comment}
+  const v1Comment = /^\/1\/(\d+)\/(?:project\/(\d+)\/)?task\/(\d+)\/comment\/(\d+)/.exec(path)
```
[the remark this answers is not in the document]
Comment (amondnet) [reply hooks/intercept-webfetch.ts:58]: @cubic-dev-ai 동일한 end-anchoring 제안을 의도적으로 받지 않았습니다(declined) — CodeRabbit의 같은 지적에 대한 응답과 동일한 이유입니다.

Asana는 실제 task URL에 focus 접미사 `/f`를 붙입니다(`/0/{project}/{task}/f`, `.../task/{task}/f`). task/comment 패턴을 `\/?$`로 정확 매칭하면 이 흔한 focus URL이 매칭에서 빠져 hook이 실제 task 인터셉트를 놓칩니다(기능 회귀). project 라우트는 이미 `$`로 앵커링되어 있고 comment는 task보다 먼저 검사되므로, task/comment의 prefix 매칭은 의도된 동작입니다.

de95547에서 V0/V1 `/f` URL → task 라우팅을 회귀 테스트로 고정해 두었습니다.
